Make AndroidView take gesture recognizer factories.

amirh opened the pull request:
Before this PR AndroidView's gestureRecognizers field was a list of
gesture recognizers. This was problematic as when the widget was rebuilt
with the same gesture recognizer instances we would try to re-join the
recognizers to a gesture arena team and crash (as a OneSeqeunceGestureRecognizer
team can only be set once).

With this change, we instead take a set of factories.
This allows AndroidView to create the gesture recognizers just before
adding them to the team, and thus be sure that they are only added once to a
team.

The factories are identified by the type of the object they create, this
allows AndroidView to know when it is given an equivalent set of gesture
recognizer factories, and do nothing in that case.

Fixes #21514.

amirh committed:
Make AndroidView take gesture recognizer factories.

Before this PR AndroidView's gestureRecognizers field was a list of
gesture recognizers. This was problematic as when the widget was rebuilt
with the same gesture recognizer instances we would try to re-join the
recognizers to a gesture arena team and crash (as a OneSeqeunceGestureRecognizer
team can only be set once).

With this change, we instead take a set of factories.
This allows AndroidView to create the gesture recognizers just before
adding them to the team, and thus be sure that they are only added once to a
team.

The factories are identified by the type of the object they create, this
allows AndroidView to know when it is given an equivalent set of gesture
recognizer factories, and do nothing in that case.

amirh commented:
Matching changes to maps and webview in flutter/plugins#778

Hixie commented on packages/flutter/lib/src/foundation/basic_types.dart:
```diff
@@ -280,3 +280,37 @@ class _LazyListIterator<E> implements Iterator<E> {
+  /// If `type` is null, [Factory.type] returns [T], else it returns `type`.
+  ///
+  /// The `constructor` parameter must not be null.
+  const Factory(this.constructor, {Type type})
```
I'm concerned about the `type` argument, especially because (as far as I can tell) we have no way to verify that it's correct. When I saw this API, I thought "that's going to be a recipe for bugs", then I searched for "type:", and the first example in this PR is, as far as I can tell, an example of a bug.

I don't really understand why we need to override the `type` though? Isn't a `Factory< VerticalDragGestureRecognizer>` a subclass of `Factory<OneSequenceGestureRecognizer>` and thus capable of being put in the same `List<Factory<OneSequenceGestureRecognizer>>` list?

Hixie commented on packages/flutter/lib/src/foundation/basic_types.dart:
```diff
@@ -280,3 +280,37 @@ class _LazyListIterator<E> implements Iterator<E> {
+
+  final Type _type;
+  /// The type of the objects created by this factory.
+  Type get type => _type ?? T;
```
nit: if we do keep this, the field should come after the getter

Hixie commented on packages/flutter/lib/src/foundation/basic_types.dart:
```diff
@@ -280,3 +280,37 @@ class _LazyListIterator<E> implements Iterator<E> {
+    }
+    final Factory<T> typedOther = other;
+    return type == typedOther.type;
+  }
```
it seems weird that two factories with different constructor would be `==`. Maybe have a separate method that compares factories for equivalence?

Hixie replied:
(oh but i guess you need this so that they match in a set?)

Hixie replied:
this is a tough one, but i think the right call here is to use a map indexed by type, and to have the factories either not have a custom ==, or to have their custom == also compare the actual constructor.

I'm worried about the confusion that one day someone would have if they did:

```dart
  var a = Factory(() => 'a');
  var b = Factory(() => 'b');
  print(a == b); // true (!)
```

Hixie replied:
ugh, but you really want it to be a set for the rest of the API...

Hixie replied:
maybe the right solution here is to make this not be a general-purpose class and just have a class specific to this purpose, with a more specific name...

Hixie replied:
or alternatively, still take a set, but then independently verify that the set only has one of each type, and match on type when the new set comes in?

Hixie commented on packages/flutter/lib/src/rendering/platform_view.dart:
```diff
@@ -61,14 +61,14 @@ class RenderAndroidView extends RenderBox {
   {
     _motionEventsDispatcher = new _MotionEventsDispatcher(globalToLocal, viewController);
-    this.gestureRecognizers = gestureRecognizers;
+    this.gestureRecognizers = gestureRecognizers ?? new Set<Factory<OneSequenceGestureRecognizer>>();
```
assuming you never mutate this set, i recommend using a static private so you can always point to the same default empty instance.

Hixie commented on packages/flutter/lib/src/rendering/platform_view.dart:
```diff
@@ -94,13 +94,13 @@ class RenderAndroidView extends RenderBox {
   /// gesture arena, the entire pointer event sequence starting from the pointer down event
   /// will be dispatched to the Android view.
-  set gestureRecognizers(List<OneSequenceGestureRecognizer> recognizers) {
+  set gestureRecognizers(Set<Factory<OneSequenceGestureRecognizer>> recognizers) {
```
the setter should be next to (immediately after) its field. Also, there should be a getter (never have a setter with no getter).

Hixie commented on packages/flutter/lib/src/rendering/platform_view.dart:
```diff
@@ -94,13 +94,13 @@ class RenderAndroidView extends RenderBox {
   /// gesture arena, the entire pointer event sequence starting from the pointer down event
   /// will be dispatched to the Android view.
-  set gestureRecognizers(List<OneSequenceGestureRecognizer> recognizers) {
+  set gestureRecognizers(Set<Factory<OneSequenceGestureRecognizer>> recognizers) {
```
I'd probably use `value` as the name of the parameter here, because it's confusing that "recognizers" is used to set "_gestureRecognizers".

Hixie replied:
(as in, it's not immediately clear when reading the code which is the field and which is the incoming parameter)

Hixie commented on packages/flutter/lib/src/widgets/platform_view.dart:
```diff
@@ -50,7 +50,7 @@ import 'framework.dart';
   /// Creates a widget that embeds an Android view.
   ///
-  /// The `viewType`, `hitTestBehavior`, and `gestureRecognizers` parameters must not be null.
+  /// The `viewType`, and `hitTestBehavior` parameters must not be null.
```
nit: first comma is now extraneous

amirh replied:
done

Hixie commented on packages/flutter/lib/src/widgets/platform_view.dart:
```diff
@@ -92,24 +91,26 @@ class AndroidView extends StatefulWidget {
   /// gesture arena, the entire pointer event sequence starting from the pointer down event
   /// will be dispatched to the Android view.
   ///
+  /// When null an empty set of gesture recognizer factories is used, and a pointer event sequence
```
strike "null"? or add "or"?

amirh replied:
Added a comma, does it make sense now?
(e.g when this is null, then an empty set is used)

Hixie commented on packages/flutter/lib/src/widgets/platform_view.dart:
```diff
@@ -92,24 +91,26 @@ class AndroidView extends StatefulWidget {
   /// will be dispatched to the Android view.
   ///
+  /// When null an empty set of gesture recognizer factories is used, and a pointer event sequence
+  /// wil only be dispatched to the Android view if no other member of the arena claimed it.
```
wil -> will

also, this whole sentence makes no sense. maybe the "and" is extraneous?

amirh replied:
rephrased a bit, is it more clear now?

Hixie commented on packages/flutter/lib/src/widgets/platform_view.dart:
```diff
@@ -118,19 +119,25 @@ class AndroidView extends StatefulWidget {
+  ///       gestureRecognizers: <Factory<OneSequenceGestureRecognizer>> [
+  ///         new Factory<OneSequenceGestureRecognizer>(
+  ///           () => new EagerGestureRecognizer(),
+  ///           type: VerticalDragGestureRecognizer,
```
type doesn't match type returned by constructor

amirh replied:
removed type

Hixie commented on packages/flutter/test/widgets/platform_view_test.dart:
```diff
@@ -562,7 +563,12 @@ void main() {
+              gestureRecognizers: <Factory<OneSequenceGestureRecognizer>> [
+                new Factory<OneSequenceGestureRecognizer>(
+                  () => new VerticalDragGestureRecognizer(),
+                  type: VerticalDragGestureRecognizer,
```
if we could make the type argument be VerticalDragGestureRecognizer, then we wouldn't need to specify the type here.

amirh replied:
done

Hixie commented on packages/flutter/test/widgets/platform_view_test.dart:
```diff
@@ -705,4 +716,24 @@ void main() {
+      gestureRecognizers: <Factory<OneSequenceGestureRecognizer>> [
+        new Factory<OneSequenceGestureRecognizer>(
+          () => new EagerGestureRecognizer(),
+          type: VerticalDragGestureRecognizer,
```
type doesn't match constructor.

amirh replied:
removed type

amirh committed:
review comments followup

amirh reviewed (commented):
Thanks for the review, PTAL

amirh committed:
don't compare factories by type

amirh commented:
Following offline discussion, removed the type based equality for factories, and added logic to compare the set of gesture recognizer types in the render object.

PTAL

Hixie commented on packages/flutter/lib/src/rendering/platform_view.dart:
```diff
@@ -94,19 +97,43 @@ class RenderAndroidView extends RenderBox {
+  ///
+  /// Setting a new set of gesture recognizer factories with different [Factory.type]s
+  /// (e.g `gestureRecognizers.map((f) => f.type).toSet()` is different, will reject any active
+  /// gesture arena.
```
this last paragraph is confusing

Hixie commented on packages/flutter/lib/src/rendering/platform_view.dart:
```diff
@@ -94,19 +97,43 @@ class RenderAndroidView extends RenderBox {
+  /// `gestureRecognizers` must not contain more than one factory with the same [Factory.type].
+  ///
+  /// Setting a new set of gesture recognizer factories with the same [Factory.type]s as the current
+  /// set is ignored.
```
s/is ignored/has no effect, because the factories' constructors would have already been called with the previous set/.

Hixie commented on packages/flutter/lib/src/rendering/platform_view.dart:
```diff
@@ -94,19 +97,43 @@ class RenderAndroidView extends RenderBox {
+  }
+
+  static Set<Type> _factoriesTypeSet<T>(Set<Factory<T>> factories) =>
+      factories.map<Type>((Factory<T> factory) => factory.type).toSet();
```
i'm amused that you used => here (even though the whole thing doesn't fit on one line), but you used `{ }` for the Factory.toString method, even though that could easily fit on one line.

Hixie commented on packages/flutter/lib/src/rendering/platform_view.dart:
```diff
@@ -211,8 +238,14 @@ class RenderAndroidView extends RenderBox {
+    team.captain = this;
+    for (OneSequenceGestureRecognizer recognizer in _gestureRecognizers) {
+      recognizer.team = team;
+    }
```
you could reduce this to one loop:

```dart
team = GestureArenaTeam();
￼team.captain = this;
_g￼estureRecognizers = gestureRecognizerFactories.map(
  (Factory<OneSequenceGestureRecognizer> factory) {
    return factory.constructor()
      ..team = team;
  }
).toSet();
```

or:

```dart
team = GestureArenaTeam();
￼team.captain = this;
_gestureRecognizers ??= new Set<OneSequenceGestureRecognizer>();
_gestureRecognizers.clear();
for (Factory<OneSequenceGestureRecognizer> factory in gestureRecognizerFactories) {
  _gestureRecognizers.add(
    factory.constructor()
      ..team = team
  );
}
```

...or some such.

amirh replied:
done

Hixie commented on packages/flutter/lib/src/widgets/platform_view.dart:
```diff
@@ -160,7 +171,7 @@ class _AndroidViewState extends State<AndroidView> {
         controller: _controller,
         hitTestBehavior: widget.hitTestBehavior,
-        gestureRecognizers: widget.gestureRecognizers,
+        gestureRecognizers: widget.gestureRecognizers ?? Set<Factory<OneSequenceGestureRecognizer>>(),
```
there's a lot of places in this PR where you either have defaults, or default from null, or whatever (sometimes with a constructor like here, sometimes with a precreated static variable, etc). It would be nice if there was only one place that needed to default, and on one side of that null was considered acceptable, and on the other side it was not.

amirh replied:
Left this the only place that defaults for null, the classes downstream now require a non null value.

Hixie commented on packages/flutter/lib/src/rendering/platform_view.dart:
```diff
@@ -94,19 +97,43 @@ class RenderAndroidView extends RenderBox {
+  /// Setting a new set of gesture recognizer factories with different [Factory.type]s
+  /// (e.g `gestureRecognizers.map((f) => f.type).toSet()` is different, will reject any active
+  /// gesture arena.
+  void setGestureRecognizers(Set<Factory<OneSequenceGestureRecognizer>> gestureRecognizers) {
```
i recommend calling this updateGestureRecognizers so that people don't tell you it should be a setter.

Hixie commented on packages/flutter/test/widgets/platform_view_test.dart:
```diff
@@ -562,7 +563,11 @@ void main() {
             child: AndroidView(
               viewType: 'webview',
-              gestureRecognizers: <OneSequenceGestureRecognizer> [VerticalDragGestureRecognizer()],
+              gestureRecognizers: <Factory<OneSequenceGestureRecognizer>> [
```
here and elsewhere: extraneous space before `[`

amirh replied:
done

Hixie commented:
![LGTM](http://junkyard.damowmow.com/579) modulo changes

amirh commented on packages/flutter/lib/src/widgets/platform_view.dart:
```diff
@@ -118,19 +119,29 @@ class AndroidView extends StatefulWidget {
   ///     child: AndroidView(
   ///       viewType: 'webview',
-  ///       gestureRecognizers: <OneSequenceGestureRecognizer>[ new VerticalDragGestureRecognizer() ],
+  ///       gestureRecognizers: <Factory<OneSequenceGestureRecognizer>> [
```
done

amirh commented on packages/flutter/lib/src/widgets/platform_view.dart:
```diff
@@ -118,19 +119,29 @@ class AndroidView extends StatefulWidget {
+  /// [EagerGestureRecognizer] is a special gesture recognizer that immediately claims the gesture
+  /// after a pointer down event.
+  ///
+  /// `gestureRecognizers` must not contain more than one factory with the same [Factory.type].
```
done

amirh committed:
review comments followup

amirh reviewed (commented):
Thanks for the thorough review!
